### PR TITLE
VideoPress: add debug() tool. Improve data-in-sync process

### LINKED ISSUE #27570: VideoPress: migrate tracks control to v6
- [x] Playback panel
- [x] Privacy and rating
- [x] Dimensions panel
- [x] Color panel
- [x] Tracks control

### PR DESCRIPTION
**Pull request by retrofox**
This PR uses the debug() tool to log actions in the dev console and also polishes and improves the video data-in-sync process.

Fixes https://github.com/Automattic/jetpack/issues/27570

# ⚙️ Data-in-sync process 

This is the name we refer to the process to keep in-sync the video data with the block through of its attributes.
Some video data fields should have the same value that block attributes, such as `title`, `description`, `privacy`, rating`, `tracks`, and so on. It means that we need to keep in-sync these data in both directions. From video data (provided by the server endpoint response body) to block attributes and vice versa.

## From data-video to block attributes

It happens when the block mounts. The app requests new video data, checks whether fields are out-of-sync, and sets the block attributes when needed.
Once the post saves, we consider block gets in sync with the video data.

### 🧪 Testing adding a new video block workflow

**Set your dev console to see the debug() lines**

* Open the dev console
* Set local storage debug item:
```
localStorage.setItem( 'debug', 'videopress:video:*' )
```

* probably, you may want to disable warns, errors, ... 
 
Screen Shot 2022-11-30 at 07 35 39

* Start with an empty post
* Add a new video block
* Upload the video
* After video uploads, confirm how the video data is propagated to the block attributes, taking a look at the dev console

image

The logs above show how the app sets the block attributes based on the new video data. 

https://user-images.githubusercontent.com/77539/204777827-320309e1-fcf0-41fa-9e62-17cf3165bdc6.mov

### 🧪 Generate an out-of-sync state

* Edit the data of the same video used before, from the VideoPress dashboard, for instance, `title`, `description`, and `privacy`

`title` and `description` | `privacy`
------|------
Screen Shot 2022-11-30 at 07 59 07 | Screen Shot 2022-11-30 at 08 00 03

* Go back to the block editor where you added the video
* Confirm after a hard refresh, the app detects these fields are out-of-sync and updates the block attributes to keep them in sync

Screen Shot 2022-11-30 at 08 05 21
 again.

Screen Shot 2022-11-30 at 08 06 45

* Save the post
* Hard refresh
* Confirm there is not any field out of sync.

## From block attributes to video data

The app watches and detects the data that needs to be updated to the video by hitting the proper endpoint. It happens when the user edits some block attributes, for instance.
The app collects all fields to update (title, description, ...) and performs a POST request right after the post saves, keeping the data in sync, but this time from the block to the video data.

### 🧪 Testing from block attributes to video-data

* Open the block sidebar
* Set `title`, `description`, ...
Screen Shot 2022-11-30 at 08 21 52

* Save the post
* Confirm the data detects these changes

image

* Confirm the app performs the POST request by taking a look at the dev console

Screen Shot 2022-11-30 at 08 23 26

* Go to the video details page of the VideoPress plugin
* Confirm the data changes you made have appropriately been propagated.

Screen Shot 2022-11-30 at 08 30 46

### 🧪 Testing auto-generated chapters

* Edit block description adding chapters
Screen Shot 2022-11-30 at 08 33 17

* Save the post
* Confirm the app detects and processes the chapters
Screen Shot 2022-11-30 at 08 33 42

* Confirm the app performs the request to upload the chapters
Screen Shot 2022-11-30 at 08 36 46

* Confirm the player re-renders
* Confirm the player has chapters
Screen Shot 2022-11-30 at 08 34 03

* Confirm the chapter is the track's control dialog
Screen Shot 2022-11-30 at 08 35 58

#### Changes proposed in this Pull Request:

* VideoPress: Add debug() tool to log actions in the dev console
* VideoPress: Improve video data-in-sync process

#### Testing instructions:

In the description

### CODE REVIEW

File: projects/packages/videopress/src/client/block-editor/hooks/use-video-data-update/index.ts [hunk 1]
```diff
@@ -213,25 +282,38 @@ export function useSyncMedia(
 				dataToUpdate?.description?.length &&
 				chapters?.length
 			) {
+				debug( 'Auto-generated chapter detected. Processing...' );
 				const track: UploadTrackDataProps = {
-					label: __( 'English', 'jetpack-videopress-pkg' ),
+					label: __( 'English (auto-generated)', 'jetpack-videopress-pkg' ),
```
Comment (dhasilva) [inline projects/packages/videopress/src/client/block-editor/hooks/use-video-data-update/index.ts:287]: :brain:

File: projects/packages/videopress/src/client/block-editor/hooks/use-video-data-update/index.ts [hunk 2]
```diff
@@ -213,25 +282,38 @@ export function useSyncMedia(
...
 					srcLang: 'en',
 					kind: 'chapters',
 					tmpFile: generateChaptersFile( dataToUpdate.description ),
 				};
 
+				debug( 'Auto-generated track: %o', track );
+
 				uploadTrackForGuid( track, guid ).then( ( src: string ) => {
+					const autoGeneratdTrackIndex = attributes.tracks.findIndex(
```
Comment (dhasilva) [inline projects/packages/videopress/src/client/block-editor/hooks/use-video-data-update/index.ts]: `autoGeneratdTrackIndex` -> `autoGeneratedTrackIndex`

File: projects/packages/videopress/src/client/block-editor/hooks/use-video-data-update/index.ts [hunk 3]
```diff
@@ -78,6 +82,51 @@ const mapFieldsToAttributes = {
 	is_private: 'isPrivate',
 };
 
+/**
+ * Re-arragne the tracks to match the block attribute format.
```
Comment (dhasilva) [inline projects/packages/videopress/src/client/block-editor/hooks/use-video-data-update/index.ts]:
```suggestion
 * Re-arrange the tracks to match the block attribute format.
```

File: projects/packages/videopress/src/client/block-editor/hooks/use-video-data-update/index.ts (hunk 2 above)
Comment (retrofox) [reply projects/packages/videopress/src/client/block-editor/hooks/use-video-data-update/index.ts]: https://github.com/Automattic/jetpack/pull/27669/commits/5cf75ea57ee68ce6c941d723e305838350c6bab7

**Review by dhasilva — APPROVED** [review]
Excellent work. Thanks!

**Pull request merged**